fc/parsing/actions.py

```diff
@@ -1200,204 +1268,353 @@ def _number_generator(self, value, units):
+        for pvar in self.protocol_variables:
+            if pvar.model_variable is self.time_variable:
+                if units is None:
+                    units = pvar.units
+                elif pvar.units is not None and pvar.units != units:
+                    raise ProtocolError(f'Inconsistent units specified for the time variable {pvar.long_name}.')
 
-            # Convert units if required
-            eq = self.units.convert_expression_recursively(eq, None)
+        # Convert if required and possible
+        if units is not None:
+            units = self.units.get_unit(units)
+            self.time_variable = self.model.convert_variable(self.time_variable, units, DataDirectionFlow.INPUT)
 
-            # Add new equation
-            self.model.add_equation(eq)
+    def _convert_units(self):
+        """Checks the units of model variables against protocol variables and converts them if needed."""
+        # Note: this code assumes that the time variable has already been converted, so no provision is made to ensure
+        # that self.time_variable still points at the correct variable, or that time is converted as an input.
 
-        # Apply all initial values set in the inputs
-        for var, value in initial_values.items():
-            if self.model.is_state(var):
-                # Set initial value
-                var.initial_value = value
-            else:
-                # Replace equation
-                old_eq = self.model.get_definition(var)
-                if old_eq is not None:
-                    self.model.remove_equation(old_eq)
-                self.model.add_equation(sympy.Eq(var, self.model.add_number(value, var.units)))
+        # TODO: This method could be extended to only do transitive-variable unit conversion if units are not given
+        #       already using a more specific annotation.
 
-        # Check for undefined local variables
-        for name, var in self.local_vars.items():
-            old_eq = self.model.get_definition(var)
-            if old_eq is None:
-                raise ProtocolError(f'Underdefined variable. A new variable {name} was created in a var statement, but'
-                                    ' no initial value or equation was assigned.')
+        # Check and convert inputs
+        for pvar in self.protocol_variables:
+            # Get desired units
+            if pvar.units is None:
+                continue
+            units = self.units.get_unit(pvar.units)
+
+            # Convert model_variable units
+            if (pvar.input or pvar.output) and pvar.model_variable is not None:
+                if pvar.model_variable.units != units:
+                    direction = DataDirectionFlow.INPUT if pvar.input else DataDirectionFlow.OUTPUT
+                    pvar.model_variable = self.model.convert_variable(pvar.model_variable, units, direction)
+
+            # Convert transitive output units
```

Comment:
I need to think through whether it's possible to do this correctly with your new `_process_protocol_variables` approach, but it's late and my brain is tired. There were definitely edge cases that meant output conversion *had* to happen after all other processing before!